neffint/fixed_grid_fourier_integral.py

```diff
@@ -0,0 +1,247 @@
+    :return: The asymptotic terms of the Fourier Integral for all times, given as an array of shape (M, X1, X2, ...)
+    :rtype: np.ndarray
+    """
+
+    # TODO: Reconsider func_value_end and func_derivative_end types, shapes and type hints
+    
+    func_value_end = np.asarray(func_value_end)
+    func_derivative_end = np.asarray(func_derivative_end)
+    
+    # Add axes from function output
+    for _ in range(func_value_end.ndim):
+        times = times[..., np.newaxis]
+    
+    # Add time axis
+    func_value_end = func_value_end[np.newaxis, ...]
+    func_derivative_end = func_derivative_end[np.newaxis, ...]
+    
+    return np.exp(1j*times*omega_end) * (1j*func_value_end/times - func_derivative_end/times**2)
+
+def fourier_integral_fixed_sampling_pchip(
+    times: ArrayLike,
+    frequencies: ArrayLike,
+    func_values: ArrayLike,
+    inf_correction_term: bool
+) -> np.ndarray:
+    """Calculates the fourier integral of the function, given as an array of values corresponding to an array of frequencies, for the time values given as input.
+    A Filon type algorithm using a piecewise cubic Hermite interpolating polynomial (pchip) and optionally an asymptotic correction term.
+    For details on implementation, see [1].
+
+    :param times: Float or 1D array of floats of length M, the time(s) to compute the fourier integral for.
+    :type times: ArrayLike
+    :param frequencies: Float or 1D array of floats of length N, the frequencies the function to be transformed have been evaluated at.
+    :type frequencies: ArrayLike
+    :param func_values: Complex or ND array of complex of shape (N, X1, X2, ...), the outputs of the function to be transformed at the frequencies given as input.
+    :type func_values: ArrayLike
+    :param inf_correction_term: True if an asymptotic correction term should be added to the output, otherwise the integral is effectively truncated at the highest frequency.
+    :type inf_correction_term: bool
+    :return: The fourier integral of the input function at the input times, given as an array of shape (M, X1, X2, ...)
+    :rtype: np.ndarray
+    
+    [1] N. Mounet. The LHC Transverse Coupled-Bunch Instability, PhD thesis 5305 (EPFL, 2012)
+    """
+
+    # Turn inputs into arrays if they are not already, switch to angular frequencies
+    omegas = 2 * np.pi * np.asarray(frequencies)
+    func_values = np.asarray(func_values)
+    times = np.asarray(times)
+    
+    # Set up result array
+    result = np.zeros((len(times), *[axis_size for axis_size in func_values.shape[1:]]), dtype=complex)
+
+    # Calculate derivatives
+    func_derivatives = complex_pchip(omegas, func_values, omegas, derivative_order=1)
+    
+    # Add asymptotic correction term
+    if inf_correction_term:
+        result += fourier_integral_inf_correction(
+            times=times,
+            omega_end=omegas[-1],
+            func_value_end=func_values[-1],
+            func_derivative_end=func_derivatives[-1]
+        )
```

Comment:
I do the asymptotic correction term first so that it happens before the array reshaping. This allows the inf_corrcetion function to do it's own reshaping, that will be handy for the adaptive algorithm